sched.c

```diff
@@ -0,0 +1,209 @@
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <ktf.h>
+#include <lib.h>
+#include <list.h>
+#include <sched.h>
+#include <string.h>
+#include <console.h>
+#include <spinlock.h>
+#include <setup.h>
+
+#include <smp/smp.h>
+
+#include <mm/vmm.h>
+
+static list_head_t tasks;
+static tid_t next_tid;
+
+static spinlock_t lock = SPINLOCK_INIT;
+
+static bool terminate;
+
+void init_tasks(void) {
+    printk("Initializing tasks\n");
+
+    list_init(&tasks);
+}
+
+static const char *task_state_names[] = {
+    [TASK_STATE_NEW]       = "NEW",
+    [TASK_STATE_READY]     = "READY",
+    [TASK_STATE_SCHEDULED] = "SCHEDULED",
+    [TASK_STATE_RUNNING]   = "RUNNING",
+    [TASK_STATE_DONE]      = "DONE",
+};
+
+static inline void set_task_state(task_t *task, task_state_t state) {
+    ASSERT(task);
+
+    dprintk("CPU[%u]: state transition %s -> %s\n", task->cpu,
+            task_state_names[task->state], task_state_names[state]);
+
+    ACCESS_ONCE(task->state) = state;
+    smp_mb();
+}
+
+static inline task_state_t get_task_state(task_t *task) {
+    task_state_t state;
+    ASSERT(task);
+
+    state = ACCESS_ONCE(task->state);
+    smp_rmb();
+
+    return state;
+}
+
+static task_t *create_task(void) {
+    task_t *task = get_free_page(GFP_KERNEL);
+
+    if (!task)
+        return NULL;
+
+    memset(task, 0, sizeof(*task));
+    task->id = next_tid++;
+    task->cpu = INVALID_CPU;
+    set_task_state(task, TASK_STATE_NEW);
+
+    spin_lock(&lock);
+    list_add(&task->list, &tasks);
+    spin_unlock(&lock);
+
+    return task;
+}
+
+static void prepare_task(task_t *task, const char *name, task_func_t func, void *arg) {
+    if (!task)
+        return;
+
+    BUG_ON(get_task_state(task) > TASK_STATE_READY);
+
+    task->name = name;
+    task->func = func;
+    task->arg = arg;
+    set_task_state(task, TASK_STATE_READY);
+}
+
+static void wait_for_task_state(task_t *task, task_state_t state) {
+    if (!task)
+        return;
+
+    while (get_task_state(task) != state)
+        cpu_relax();
+}
+
+task_t *new_task(const char *name, task_func_t func, void *arg) {
+    task_t *task = create_task();
+
+    if (!task)
+        return NULL;
```

Comment:
Nitpick: redundant check

Reply:
I would keep it for the just-in-case sake. Thanks.